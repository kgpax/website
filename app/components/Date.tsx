import dayjs from 'dayjs'
import advancedFormat from 'dayjs/plugin/advancedFormat'

dayjs.extend(advancedFormat)

type DateProps = React.HTMLAttributes<HTMLElement> & {
  children?: string
}

export default function Date({ children }: DateProps) {
  const formattedDate = children
    ? dayjs(children).format('Do MMM YYYY')
    : undefined
  return children ? (
    <time dateTime={children} className="opacity-70">
      - {formattedDate} -
    </time>
  ) : (
    <>Publish date TBC</>
  )
}
