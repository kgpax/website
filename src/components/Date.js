import styled from 'styled-components';

const StyledDate = styled.time`
  display: block;
  text-align: center;
  opacity: 0.7;
  &::before {
    content: '- ';
  }
  &::after {
    content: ' -';
  }
`;

const Date = (props) => <StyledDate {...props} />;

export default Date;
