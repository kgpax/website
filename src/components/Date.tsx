import dayjs from 'dayjs'
import advancedFormat from 'dayjs/plugin/advancedFormat'

dayjs.extend(advancedFormat)

type DateProps = React.HTMLAttributes<HTMLElement> & {
  children?: string
}

export default function Date({ children, ...props }: DateProps) {
  const formattedDate = children
    ? dayjs(children).format('Do MMM YYYY')
    : undefined
  return children ? (
    <time dateTime={children} {...props}>
      {formattedDate}
    </time>
  ) : (
    <span {...props}>Publish date TBC</span>
  )
}
