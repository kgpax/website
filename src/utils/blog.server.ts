import glob from 'fast-glob'
import {
  BlogData,
  BlogEntry,
  BlogSeriesEntry,
  EntryMetaData,
} from '~/types/blog'

export async function getBlogEntries(): Promise<BlogEntry[]> {
  const postPaths = await glob('src/pages/blog/**/*.mdx')

  const posts = await Promise.all(
    postPaths.map(async path => {
      const filePath = path.replace(/^src\/pages/, '')
      const fileName = path.split('/').reverse()[0].replace(/\.mdx/, '')
      const slug = filePath.replace(/\.mdx/, '')
      const series = filePath
        .substring(0, filePath.lastIndexOf('/'))
        .replace(/^\/blog\//, '')
      const { meta }: { meta: EntryMetaData } = await import(
        `../pages${filePath}`
      )

      return {
        ...meta,
        filePath,
        fileName,
        series,
        slug,
        isPublished: !!meta.date,
      } as BlogData
    }),
  )

  const sortedPosts = posts.sort((a, b) =>
    (a.date ?? '9999-99-99').localeCompare(b.date ?? '9999-99-99'),
  )

  const entries: BlogEntry[] = []

  for (const post of sortedPosts) {
    if (!post.series) {
      entries.push({
        type: 'post',
        ...post,
      })
    } else {
      if (post.series) {
        let series = entries.find(
          entry => entry.series === post.series,
        ) as BlogSeriesEntry
        if (!series) {
          series = {
            type: 'series',
            ...post,
            posts: [],
          }
          entries.push(series)
        }
        if (post.fileName !== '_meta') {
          series.posts.push({ type: 'post', ...post })
        }
      } else {
      }
    }
  }

  return entries
}
