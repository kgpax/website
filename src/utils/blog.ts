import { BlogSeriesEntry } from '~/types/blog'

export function getBlogPostSeries(
  series: BlogSeriesEntry[],
  id: string,
): BlogSeriesEntry | null {
  return series.find(x => x.posts.some(y => y.id === id)) ?? null
}
