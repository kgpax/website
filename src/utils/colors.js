const pairs = [
  ['#D7F2BA', '#BDE4A8'],
  ['#FFF200', '#FFD300'],
  ['#97D2FB', '#83BCFF'],
  ['#FFBE85', '#FF9B42'],
];

export const pickColorPair = () =>
  pairs[Math.floor(Math.random() * pairs.length)];
