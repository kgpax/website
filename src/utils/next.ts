import { GetStaticPropsResult } from 'next'

export function next404<T>(): GetStaticPropsResult<T> {
  return { notFound: true }
}
