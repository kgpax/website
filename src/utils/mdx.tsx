import { MDXComponents } from 'mdx/types'
import { BlogPage, Heading, Link, MdxCode } from '~/components'

export const components: MDXComponents = {
  wrapper: props => {
    console.log(props)
    return <BlogPage {...props}>{props.children}</BlogPage>
  },
  h2: Heading,
  a: Link,
  code: MdxCode,
}
