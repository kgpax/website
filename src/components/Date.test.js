import { render } from '@test/utils';
import Date from '@components/Date';

const component = (props = {}) =>
  render(<Date {...props}>27th August 2021</Date>);

test('should match snapshot', () => {
  expect(component().element).toMatchSnapshot();
});

test('should be a TIME element', () => {
  const { element } = component();
  expect(element.tagName).toBe('TIME');
});

test('should contain expected content', () => {
  const { element } = component();
  expect(element.textContent).toBe('27th August 2021');
});
