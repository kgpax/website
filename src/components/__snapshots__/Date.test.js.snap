// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<time
  class="Date__StyledDate-sc-mc7f4f-0 lnxnLg"
>
  27th August 2021
</time>
`;
